File: app/src/main/java/net/tjado/usbgadget/GadgetObject.java
package net.tjado.usbgadget;

import java.util.ArrayList;
import java.util.HashMap;

public class GadgetObject {
    protected HashMap<String, String> values = new HashMap<>();
    protected ArrayList<String> functions = new ArrayList<>();
    protected ArrayList<String> activeFunctions = new ArrayList<>();
    protected GadgetShellApi gsa = new GadgetShellApi();

    public GadgetObject() {
    }

    public void setValue(String key, String value) {
        values.put(key, value);
    }

    public String getValue(String key) {
        try {
            return values.get(key);
        }
        catch (Exception e) {
            return "not set";
        }
    }

    public void addFunction(String f) {
        functions.add(f);
    }

    public ArrayList<String> getFunctions() {
        return functions;
    }

    public void addActiveFunction(String f) {
        activeFunctions.add(f);
    }

    public ArrayList<String> getActiveFunctions() {
        return activeFunctions;
    }

    public Boolean isActivated() {
        String udc = getValue("udc");
        return !udc.equals("") && !udc.equals("not set");
    }

    public String getFormattedFunctions() {
        StringBuffer sb =  new StringBuffer();
        String color = "#ff0000";

        for(String function : functions) {
            if( activeFunctions.contains(function)) {
                color = "#008000";
            }
            sb.append(String.format("<font color=%s>%s</font><br />", color, function));
        }

        return sb.toString();
    }

    public Boolean activate() {
        return activate(null);
    }

    public Boolean activate(RootTask.OnRootTaskListener onRootTaskFinished) {
        return this.gsa.activate(getValue("gadget_path"), onRootTaskFinished);
    }

    public Boolean deactivate() {
        return deactivate(null);
    }

    public Boolean deactivate(RootTask.OnRootTaskListener onRootTaskFinished) {
        return this.gsa.deactivate(getValue("gadget_path"), onRootTaskFinished);
    }


    public Boolean activateFunction(String function, Boolean reactivateGadget) {
        String gadgetPath = getValue("gadget_path");
        String gadgetConfigPath = getValue("config_path");

        Boolean status = this.gsa.activateFunction(gadgetPath, gadgetConfigPath, function, (isActivated() && reactivateGadget));
        if (status) {
            // add activated function to activeFunctions to avoid refreshing configFS data
            activeFunctions.add(function);
        }

        return status;
    }

    public Boolean deactivateFunction(String function, Boolean reactivateGadget) {
        String gadgetPath = getValue("gadget_path");
        String gadgetConfigPath = getValue("config_path");

        Boolean status = this.gsa.deactivateFunction(gadgetPath, gadgetConfigPath, function, (isActivated() && reactivateGadget));
        if (status) {
            // remove deactivated function from activeFunctions to avoid refreshing configFS data
            activeFunctions.removeIf(function::equals);
        }

        return status;
    }
}

File: app/src/main/java/net/tjado/usbgadget/GadgetObject.kt
package net.tjado.usbgadget

import net.tjado.usbgadget.RootTask.OnRootTaskListener


class GadgetObject {
    var values = HashMap<String, String>()
    var functions = ArrayList<String>()
        private set
    var activeFunctions = ArrayList<String>()
        private set
    var shellApi = GadgetShellApi()

    fun setValue(key: String, value: String) {
        values[key] = value
    }

    fun getValue(key: String): String {
        return try {
            values.getValue(key)
        } catch (e: Exception) {
            "not set"
        }
    }

    fun addFunction(f: String) {
        functions.add(f)
    }

    fun addActiveFunction(f: String) {
        activeFunctions.add(f)
    }

    val isActivated: Boolean
        get() {
            val udc = getValue("udc")
            return udc != "" && udc != "not set"
        }

    val formattedFunctions: String
        get() {
            val sb = StringBuffer()
            for (function in functions) {
                var color = "#ff0000"
                if (activeFunctions.contains(function)) {
                    color = "#008000"
                }
                sb.append(String.format("<font color=%s>%s</font><br />", color, function))
            }
            return sb.toString()
        }

    fun activate(onRootTaskFinished: OnRootTaskListener? = null): Boolean {
        return shellApi.activate(getValue("gadget_path"), onRootTaskFinished)
    }

    fun deactivate(onRootTaskFinished: OnRootTaskListener? = null): Boolean {
        return shellApi.deactivate(getValue("gadget_path"), onRootTaskFinished)
    }

    fun activateFunction(function: String, reactivateGadget: Boolean): Boolean {
        val gadgetPath = getValue("gadget_path")
        val gadgetConfigPath = getValue("config_path")

        val status = shellApi.activateFunction(
            gadgetPath,
            gadgetConfigPath,
            function,
            isActivated && reactivateGadget
        )

        if (status) {
            // add activated function to activeFunctions to avoid refreshing configFS data
            activeFunctions.add(function)
        }

        return status
    }

    fun deactivateFunction(function: String, reactivateGadget: Boolean): Boolean {
        val gadgetPath = getValue("gadget_path")
        val gadgetConfigPath = getValue("config_path")

        val status = shellApi.deactivateFunction(
            gadgetPath,
            gadgetConfigPath,
            function,
            isActivated && reactivateGadget
        )

        if (status) {
            // remove deactivated function from activeFunctions to avoid refreshing configFS data
            activeFunctions.removeIf { anObject -> function == anObject }
        }

        return status
    }
}